### PR TITLE
(cloudwatch) expand multi select template variable

### PR DESCRIPTION
**Pull request by mtanda**
reopen https://github.com/grafana/grafana/pull/4421 for development branch.

### CODE REVIEW

**Comment by dpost** [comment]
Any update on when this will be pulled in? It'd be a great feature to have

**Comment by niclic** [comment]
+1 This will be very useful.

**Comment by damienpuig** [comment]
+1 Would be awesome!

**Comment by mtanda** [comment]
rebase and change branch destination.

**Comment by mgobec** [comment]
@mtanda Awesome job with this. Really needed for any aws deployment.

**Comment by mtanda** [comment]
@mgobec Hi, do you success to use multiple select?

File: public/app/plugins/datasource/cloudwatch/datasource.js [hunk 1]
```diff
@@ -332,6 +333,36 @@ function (angular, _, moment, dateMath, CloudWatchAnnotationQuery) {
       });
     }
 
+    this.expandTemplateVariable = function(targets, templateSrv) {
```
Comment (bergquist) [inline public/app/plugins/datasource/cloudwatch/datasource.js]: Would be nice to splitt this functions into smaller functions. Makes it easier to read the code.

Comment (mtanda) [inline public/app/plugins/datasource/cloudwatch/datasource.js]: I simplify the code.

**Comment by Esya** [comment]
Any update on this?

**Comment by mtanda** [comment]
rebase again.

**Comment by bergquist** [comment]
Thank you for contributing!

**Comment by mtanda** [comment]
Yay! Thanks!!

**Comment by paco3346** [comment]
`public_gen/app/features/templating/templateSrv.js` does not contain a `containsVariable` method which causes a 'is not a function' error to be thrown
```
TypeError: templateSrv.containsVariable is not a function
    at eval (http://localhost:3001/public/app/plugins/datasource/cloudwatch/datasource.js?bust=1476464369366:382:32)
    at baseFindIndex (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:808:11)
    at findIndex (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:6898:14)
    at Function.eval (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:4892:21)
    at eval (http://localhost:3001/public/app/plugins/datasource/cloudwatch/datasource.js?bust=1476464369366:381:28)
    at arrayMap (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:645:23)
    at Function.map (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:9227:14)
    at interceptor (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:16619:35)
    at thru (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:8459:14)
    at eval (http://localhost:3001/public/vendor/lodash/dist/lodash.js?bust=1476464369366:4216:28)
```

**Comment by mpareja** [comment]
We ran into that too and found a number of other issues with the current implementation:

1. The "All" multi-selection template variable functionality doesn't work
2. Having more than one template variable breaks things

Both those issues should be fixable in `datasource/cloudwatch/datasource.js`.

Problem 1 can be resolved by looking for the selected All option and generating the `options.targets` as if all other options were selected. Here's a dirty hack that proves the functionality is feasible - I doubt it's implemented in the _Grafana way of doing things_:

```
      const modifiedTemplateSrv = _.cloneDeep(templateSrv);
      modifiedTemplateSrv.variables
        .filter(v => v.options.some(o => o.selected && o.value === '$__all'))
        .forEach(v => v.options.forEach(o => o.selected = !o.selected))
```

Problem 2 is pretty clear when you consider what the new function added in this PR is named: `expandTemplateVariable`. The function attempts to do the job for a single template variable (the first one found by the `_.findKey` on [line 357](https://github.com/grafana/grafana/pull/5003/files#diff-a8259c8cbce476f0faab44ef81675dfbR357)). It should be doing it for all template variables which could include other "All" selections.

**Comment by paco3346** [comment]
@mpareja I think grafana calls `expandTemplateVariable` per variable. I'm able to run multiple graphs with multiple variables and everything works as expected.

**Comment by mpareja** [comment]
@paco3346 Are you able to run a single graph that references 2 different template variables in the dimensions?

**Pull request merged**